src/sinks/s3_common/sink.rs

```diff
@@ -0,0 +1,305 @@
+use crate::sinks::util::sink::ServiceLogic;
+use crate::{
+    config::{log_schema, SinkContext},
+    event::Event,
+    sinks::util::{
+        buffer::GZIP_FAST, encoding::EncodingConfiguration, sink::StdServiceLogic, Compression,
+    },
+};
+use async_trait::async_trait;
+use bytes::Bytes;
+use chrono::Utc;
+use flate2::write::GzEncoder;
+use futures::{
+    stream::{BoxStream, FuturesUnordered, StreamExt},
+    FutureExt, TryFutureExt,
+};
+use std::{
+    collections::HashMap,
+    convert::TryInto,
+    fmt::Debug,
+    io::{self, Write},
+    num::NonZeroUsize,
+    time::Duration,
+};
+use tokio::{
+    pin, select,
+    sync::{
+        mpsc::{channel, Receiver},
+        oneshot,
+    },
+};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+use uuid::Uuid;
+use vector_core::{
+    buffers::Acker,
+    event::{EventFinalizers, Finalizable},
+    sink::StreamSink,
+    stream::batcher::Batcher,
+};
+
+use crate::sinks::s3_common::partitioner::KeyPartitioner;
+use crate::sinks::s3_common::service::S3Request;
+use crate::sinks::util::encoding::EncodingConfig;
+use crate::sinks::util::sink::Response;
+
+pub struct S3Sink<S, R>
+where
+    R: S3RequestBuilder,
+{
+    acker: Option<Acker>,
+    service: Option<S>,
+    request_builder: R,
+    partitioner: Option<KeyPartitioner>,
+    batch_size_bytes: Option<NonZeroUsize>,
+    batch_size_events: NonZeroUsize,
+    batch_timeout: Duration,
+}
+
+impl<S, R> S3Sink<S, R>
+where
+    R: S3RequestBuilder,
+{
+    pub fn new(
+        cx: SinkContext,
+        service: S,
+        request_builder: R,
+        partitioner: KeyPartitioner,
+        batch_size_bytes: Option<NonZeroUsize>,
+        batch_size_events: NonZeroUsize,
+        batch_timeout: Duration,
+    ) -> Self {
+        Self {
+            acker: Some(cx.acker()),
+            service: Some(service),
+            request_builder,
+            partitioner: Some(partitioner),
+            batch_size_bytes,
+            batch_size_events,
+            batch_timeout,
+        }
+    }
+}
+
+#[async_trait]
+impl<S, R> StreamSink for S3Sink<S, R>
```

Comment:
I'm curious what @tobz thinks here. I like the lift of the config bits but along with #8825 we've got three very similar sink implementations going. My preference would be to duplicate the sink and run_io etc into each sink module so those are allowed to vary as needed until we come up with some higher-level abstraction, once we have some notion about what common pieces are in play.

Reply:
I see, yeah, I'm ok with both options, indeed until there is a clear vision of common pieces, we could duplicate this logic. Waiting for @tobz's input on that.

Reply:
My gut feeling here is really that I'm not afraid to move it all back out, after this PR, if it turns out that abstracting it was the wrong move.

If it currently works, I'm fine with it.

Reply:
I have a feeling(but you both might have a better understanding of the current state) that soon-ish we might want to pull up common interfaces like this, so maybe it would be less work in this case. But as @tobz mentioned, it is not too much work if we decide to copy it either. So I'd probably keep it as is.